### PR TITLE
In Sys.getcwd, don't call getwd if HAS_GETWD is not set (follow-up to #1086)

### PR DESCRIPTION
**Pull request by xavierleroy**
This is a proper fix for the bug reported in #1086 but incorrectly fixed there.

### CODE REVIEW

File: byterun/sys.c [hunk 1]
```diff
@@ -336,11 +336,13 @@ CAMLprim value caml_sys_chdir(value dirname)
 CAMLprim value caml_sys_getcwd(value unit)
 {
   char buff[4096];
```
Comment (yallop) [inline byterun/sys.c]: This is perhaps slightly off-topic, but shouldn't the length of `buff` use `PATH_MAX` rather than `4096`?

Comment (xavierleroy) [reply byterun/sys.c]: Well, for modern OSes and file systems, there is no fixed limit on the length of an absolute file path, so 4096 is not much more wrong than `PATH_MAX`...

We probably could do a better job by dynamically resizing the buffer if `getcwd` errors with `ERANGE`.  But that's a topic for another pull request.

**Comment by alainfrisch** [comment]
Copying a comment from @gasche on the other PR:

> there is another occurrence of getwd in asmrun/spacetime.c that you may want to change as well

**Comment by shindere** [comment]
Xavier Leroy (2017/09/14 18:00 +0000):
> xavierleroy commented on this pull request.
> 
> 
> 
> > @@ -336,11 +336,13 @@ CAMLprim value caml_sys_chdir(value dirname)
>  CAMLprim value caml_sys_getcwd(value unit)
>  {
>    char buff[4096];
> 
> Well, for modern OSes and file systems, there is no fixed limit on the
> length of an absolute file path, so 4096 is not much more wrong than
> `PATH_MAX`...

Perhaps not much more wrong, but it carries less semeantics.

Reminds me the quote saying that programs are written mainly to be read,
not to be compiled. :)

**Comment by damiendoligez** [comment]
cc @mshinwell on Gabriel's note about `spacetime.c`.

**Comment by mshinwell** [comment]
I believe I have fixed `spacetime.c` directly on this branch.
I also merged with current trunk.

The `PATH_MAX` issue is a potential security concern I think (depending on the no doubt many and various implementations of `getwd`).

**Comment by xavierleroy** [comment]
@mshinwell is correct that `getwd` is a security risk.  So, let's stop using it entirely  (commit [01c90f5]).  Quoting the man page:
```
       getwd() is present in POSIX.1-2001, but  marked  LEGACY.   POSIX.1-2008
       removes   the   specification   of   getwd().   Use  getcwd()  instead.
       POSIX.1-2001 does not define any errors for getwd().
```
Even the Microsoft CRT provides `getcwd`, so that should be good enough for everyone.

I stand by my decision to not use `PATH_MAX`.  Quoting POSIX.1-2008:
> A definition of one of the symbolic constants in the following list [including `PATH_MAX`] shall be omitted from the <limits.h> header on specific implementations where the corresponding value is equal to or greater than the stated minimum, but where the value can vary depending on the file to which it is applied. The actual value supported for a specific pathname shall be provided by the pathconf() function.

So, not only `PATH_MAX` may not be a constant, but it may not be defined at all.  Enough.

**Comment by damiendoligez** [comment]
@mshinwell you should also avoid `PATH_MAX` in the spacetime code.

**Comment by xavierleroy** [comment]
I took the liberty of reverting the Spacetime code to use a fixed-size buffer and forget about PATH_MAX.  I also rebased on current trunk.  Requesting final review and permission to merge.

File: Changes [hunk 2]
```diff
@@ -609,6 +609,9 @@ Release branch for 4.06:
 - GPR#1073: Remove statically allocated compare stack.
   (Stephen Dolan)
 
+- GPR#1086: in Sys.getcwd, don't call getwd() if HAS_GETWD is not set
+  (report and first fix by Sebastian Markbåge, final fix by Xavier Leroy)
```
Comment (mshinwell) [inline Changes]: This needs updating to reflect the fact that we no longer use `getwd` at all.

File: asmrun/spacetime.c [hunk 3]
```diff
@@ -225,17 +225,15 @@ void caml_spacetime_initialize(void)
...
         }
 #else
-        if (getwd(cwd) == NULL) {
-          dir_ok = 0;
-        }
+        dir_ok = 0;
 #endif
         if (dir_ok) {
-          automatic_snapshot_dir = strdup(cwd);
+          automatic_snapshot_dir = cwd;
```
Comment (mshinwell) [inline asmrun/spacetime.c]: This is wrong, we still need `strdup`, since `cwd` is on the stack.

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Comment by mshinwell** [comment]
OK to merge after the two changes I just pointed out.

File: Changes (hunk 2 above)
Comment (xavierleroy) [reply Changes]: Done.

File: asmrun/spacetime.c (hunk 3 above)
Comment (xavierleroy) [reply asmrun/spacetime.c]: Well spotted, sorry about that.  Fixed.

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
Merged to trunk and cherry-picked to 4.06 (`538e110ad`).

**Pull request merged**